modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.uhndata.cards.internal;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.jackrabbit.oak.api.CommitFailedException;
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.spi.commit.DefaultValidator;
+import org.apache.jackrabbit.oak.spi.commit.Validator;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.jetbrains.annotations.Nullable;
+
+public class RequiredSubjectTypesValidator extends DefaultValidator
+{
+    private final ResourceResolverFactory rrf;
+
+    public RequiredSubjectTypesValidator(ResourceResolverFactory rrf)
+    {
+        this.rrf = rrf;
+    }
+
+    @Override
+    public @Nullable Validator childNodeAdded(String name, NodeState after) throws CommitFailedException
+    {
+        // Get the type of this node. Return immediately if it's not a cards:Form node
+        final String childNodeType = after.getName("jcr:primaryType");
+        if (!("cards:Form".equals(childNodeType))) {
+            return this;
+        }
+
+        // Get the jcr:uuid values for the Form's associated Questionnaire and Subject
+        final String questionnaireUUID = after.getProperty("questionnaire").getValue(Type.REFERENCE).toString();
+        final String subjectUUID = after.getProperty("subject").getValue(Type.REFERENCE).toString();
+        // Obtain a ResourceResolver for querying the JCR
+        final Map<String, Object> parameters =
+            Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, "requiredSubjectTypesValidator");
+        try (ResourceResolver serviceResolver = this.rrf.getServiceResourceResolver(parameters)) {
+
+            // Get the Subject Resource associated with this Form
+            final Resource subject = getSubjectResourceByUuid(serviceResolver, subjectUUID);
+            if (subject == null) {
+                return this;
+            }
+
+            // Get the type value for the Subject
+            final String type = subject.getValueMap().get("type", "");
+            // Get the Questionnaire Resource associated with this Form
+            final Resource questionnaire = getQuestionnaireResourceByUuid(serviceResolver, questionnaireUUID);
+            if (questionnaire == null) {
+                return this;
+            }
+
+            // Get the requiredSubjectTypes value for the Questionnaire
+            final List<String> allRequiredSubjectTypes = questionnaire.getValueMap().get("requiredSubjectTypes",
+                    List.of());
```

Comment:
This will always return an empty `List<String>` and therefore will always cause the creation of a new Form to fail. You would want to do something like

```suggestion
            final List<String> allRequiredSubjectTypes = Arrays.asList(
                questionnaire.getValueMap().get("requiredSubjectTypes", String[].class)
            );
```

instead - be sure to `import java.util.Arrays`.